### PR TITLE
General XPath fixes for the Python3 runtime

### PR DESCRIPTION
**Pull request by JohnnyonFlame**
This commit:
- Fixes some code that does not reflect the current API
- Fixes wrong XPath library behavior, such as failing to apply the "invert" flag on certain operations
- Does general style touch-ups

Further work is necessary, such as implementing tests for this functionality, so that further changes to the API or Runtime won't break functionality, and maybe separate the interpreter data from the library code. For now, a code review would be stellar.

### CODE REVIEW

File: runtime/Python3/src/antlr4/xpath/XPath.py [hunk 1]
```diff
@@ -210,24 +210,32 @@ def recover(self, e):
...
+                ttype = (tsource.ruleNames.index(word) + 1) if word in tsource.ruleNames    else None
+            else:
+                litType = tsource.literalNames.index(word)  if word in tsource.literalNames else None
+
+            # Decide which one is it
+            ttype = ttype or litType
 
-            if ttype==Token.INVALID_TYPE:
-                raise Exception( word + " at index " + str(wordToken.startIndex) + " isn't a valid token name")
+            if not ttype:
```
Comment (ericvergnaud) [inline runtime/Python3/src/antlr4/xpath/XPath.py]: mmm... this is javascript style not python, please keep the intent explicit

File: runtime/Python3/src/antlr4/xpath/XPath.py [hunk 2]
```diff
@@ -246,18 +254,16 @@ def evaluate(self, t:ParseTree):
...
 
         work = [dummyRoot]
-
-        for i in range(0, len(self.elements)):
-            next = set()
+        for element in self.elements:
+            work_next = list()
             for node in work:
-                if len( node.children) > 0 :
+                if not isinstance(node, TerminalNode) and node.children:
```
Comment (ericvergnaud) [inline runtime/Python3/src/antlr4/xpath/XPath.py]: That is probably wrong,g, not sure there is guarantee that node.children is never empty

File: runtime/Python3/src/antlr4/xpath/XPath.py [hunk 3]
```diff
@@ -210,24 +210,32 @@ def recover(self, e):
...
+            tsource = self.parser.getTokenStream().tokenSource
+
+            ttype = litType = None
+            if wordToken.type == XPathLexer.TOKEN_REF:
+                ttype = (tsource.ruleNames.index(word) + 1) if word in tsource.ruleNames    else None
+            else:
+                litType = tsource.literalNames.index(word)  if word in tsource.literalNames else None
+
+            # Decide which one is it
+            ttype = ttype or litType
```
Comment (ericvergnaud) [inline runtime/Python3/src/antlr4/xpath/XPath.py]: mmm... this is javascript style not python, please keep the intent explicit
also in Python 1 variable is probably sufficient

File: runtime/Python3/src/antlr4/xpath/XPath.py [hunk 4]
```diff
@@ -210,24 +210,32 @@ def recover(self, e):
...
+            if not ttype:
+                raise Exception("%s at index %d isn't a valid token name" % (word, wordToken.tokenIndex))
             return XPathTokenAnywhereElement(word, ttype) if anywhere else XPathTokenElement(word, ttype)
 
         else:
+            ruleIndex = self.parser.ruleNames.index(word) if word in self.parser.ruleNames else None
 
-            if ruleIndex==-1:
-                raise Exception( word + " at index " + str(wordToken.getStartIndex()) + " isn't a valid rule name")
+            if not ruleIndex:
```
Comment (ericvergnaud) [inline runtime/Python3/src/antlr4/xpath/XPath.py]: same here

**Comment by ericvergnaud** [comment]
Hi,
thanks for this
Any chance you could provide some tests that can help validate this PR?

File: runtime/Python3/src/antlr4/xpath/XPath.py (hunk 2 above)
Comment (JohnnyonFlame) [reply runtime/Python3/src/antlr4/xpath/XPath.py]: https://www.python.org/dev/peps/pep-0008/
>For sequences, (strings, lists, tuples), use the fact that empty sequences are false.
>
>```Python3
>Yes: if not seq:
>     if seq:
>
>No:  if len(seq):
>     if not len(seq):
>```

**Comment by JohnnyonFlame** [comment]
@ericvergnaud 
First, thank you for the review.
I am going to port some of the test of the Java runtime when possible. I was first waiting for a code review while taking a look at some other problems.

And yeah, the implicit intent is a bad habit, _oops_.

**Comment by JohnnyonFlame** [comment]
@ericvergnaud Implemented some of the recommended changes, corrected some bugs and included a port of _testValidPaths_ from "_tool-testsuite/test/org/antlr/v4/test/tool/TestXPath.java_"

File: runtime/Python3/src/antlr4/xpath/XPath.py [hunk 5]
```diff
@@ -210,24 +210,31 @@ def recover(self, e):
...
 
+        word = wordToken.text
         if wordToken.type==XPathLexer.WILDCARD :
-
             return XPathWildcardAnywhereElement() if anywhere else XPathWildcardElement()
 
         elif wordToken.type in [XPathLexer.TOKEN_REF, XPathLexer.STRING]:
+            tsource = self.parser.getTokenStream().tokenSource
+
+            ttype = -1
```
Comment (ericvergnaud) [inline runtime/Python3/src/antlr4/xpath/XPath.py]: I would set type to Token.INVALID_TYPE

File: runtime/Python3/src/antlr4/xpath/XPath.py [hunk 6]
```diff
@@ -210,24 +210,31 @@ def recover(self, e):
...
+            if wordToken.type == XPathLexer.TOKEN_REF:
+                if word in tsource.ruleNames:
+                    ttype = tsource.ruleNames.index(word) + 1
+            else:
+                if word in tsource.literalNames:
+                    ttype = tsource.literalNames.index(word)
 
-            if ttype==Token.INVALID_TYPE:
-                raise Exception( word + " at index " + str(wordToken.startIndex) + " isn't a valid token name")
+            if ttype == -1:
```
Comment (ericvergnaud) [inline runtime/Python3/src/antlr4/xpath/XPath.py]: such that you can test against it
not sure tsource... call may or not return it

File: runtime/Python3/src/antlr4/xpath/XPath.py (hunk 2 above)
Comment (ericvergnaud) [reply runtime/Python3/src/antlr4/xpath/XPath.py]: Not sure I support that pep...

**Comment by ericvergnaud** [comment]
Thanks for this
Apart from 2 minor style comments it looks good!

File: runtime/Python3/src/antlr4/xpath/XPath.py (hunk 6 above)
Comment (JohnnyonFlame) [reply runtime/Python3/src/antlr4/xpath/XPath.py]: Passes the tests it seems, not too sure what would be a better choice.

**Comment by JohnnyonFlame** [comment]
Thank you. Included the proposed style changes.

**Comment by ericvergnaud** [comment]
Cool!
@parrt blessed

**Pull request merged**